pkg/plugin/instance/libvirt/instance_test.go

```diff
@@ -0,0 +1,139 @@
+
+	for _, d := range doms {
+		info, err := d.GetInfo()
+		require.NoError(t, err)
+
+		if info.State != libvirt.DOMAIN_RUNNING {
+			continue
+		}
+
+		name, err := d.GetName()
+		require.NoError(t, err)
+		if name == id {
+			return true
+		}
+	}
+	return false
+}
+
+func findInstance(t *testing.T, plugin instance.Plugin, id instance.ID) *instance.Description {
+	descrs, err := plugin.DescribeInstances(map[string]string{}, false)
+	require.NoError(t, err)
+
+	for _, d := range descrs {
+		if d.ID == id {
+			return &d
+		}
+	}
+	return nil
+}
+
+// checks if an instance exists from the plugin's PoV
+func instanceExists(t *testing.T, plugin instance.Plugin, id instance.ID) bool {
+	return findInstance(t, plugin, id) != nil
+}
+
+func TestBasicLifecycle(t *testing.T) {
+	conn, err := libvirt.NewConnect(libvirtDrvTest)
+	require.NoError(t, err)
+	defer conn.Close()
+
+	plugin := NewLibvirtPlugin(libvirtDrvTest)
+	require.NotNil(t, plugin)
+
+	props := types.Any(`{
+    "DomainType": "test",
+    "Kernel": "/foo/bar/Image",
+    "CPUs": 1,
+    "Memory" : 1024
+}`)
+
+	logicalID := instance.LogicalID("spock")
+
+	id, err := plugin.Provision(instance.Spec{
+		Properties:  &props,
+		Tags:        map[string]string{},
+		Init:        "",
+		LogicalID:   &logicalID,
+		Attachments: []instance.Attachment{},
+	})
+	require.NoError(t, err)
+	require.NotNil(t, id)
+
+	require.True(t, domainExists(t, conn, string(*id)), "domain was not found in libvirt domains")
+	require.True(t, instanceExists(t, plugin, *id), "domain was not found in plugin instances")
+
+	inst := findInstance(t, plugin, *id)
+	require.NotNil(t, inst)
+
+	require.Equal(t, logicalID, *inst.LogicalID)
+
+	err = plugin.Destroy(*id)
+	require.NoError(t, err)
+
+	require.False(t, domainExists(t, conn, string(*id)), "domain was found in libvirt domains after destroy")
+	require.False(t, instanceExists(t, plugin, *id), "domain was found in plugin instances after destroy")
+}
+
+func TestLabel(t *testing.T) {
+	conn, err := libvirt.NewConnect(libvirtDrvTest)
+	require.NoError(t, err)
+	defer conn.Close()
+
+	domname := "test-instanceXXXX"
+
+	domcfg := libvirtxml.Domain{
+		Type:   "test",
+		VCPU:   &libvirtxml.DomainVCPU{Value: 1},
+		Memory: &libvirtxml.DomainMemory{Value: 1024, Unit: "MiB"},
+		Name:   "test-instanceXXXX",
+		OS: &libvirtxml.DomainOS{
+			Type: &libvirtxml.DomainOSType{
+				Type: "hvm",
+			},
+			Kernel:     "/foo/bar/Kernel",
+			Initrd:     "/foo/bar/Ramdisk",
+			KernelArgs: "root=/dev/sda2",
+		},
+	}
+
+	xmldoc, err := domcfg.Marshal()
+	require.NoError(t, err)
+
+	_, err = conn.DomainCreateXML(string(xmldoc), 0)
+	require.NoError(t, err)
+
+	require.True(t, domainExists(t, conn, domname), "domain was not found in libvirt domains")
+
+	plugin := NewLibvirtPlugin(libvirtDrvTest)
+	require.NotNil(t, plugin)
+
+	require.False(t, instanceExists(t, plugin, instance.ID(domname)), "domain was found in plugin instances before label")
+
+	labels := map[string]string{}
```

Comment:
>  Set some label when provisioning and add or edit labels with Label function.

I'll do that when someone points me to the documentation regarding the correct/expected semantics of the `Label` method. Otherwise I'm just guessing how it supposed to work and writing the answers to my guesses in both the implementation and the test.